scripts/appbuilder

```diff
@@ -1,62 +1,151 @@
+            ;;
+    esac
+done
+
+# defaults
+IMAGE_NAME=${IMAGE_NAME:-}
+OUTPUT_NAME=${OUTPUT_NAME:-appdir}
+DELETE_OUTPUT=${DELETE_OUTPUT:-false}
+OUTPUT_FORMAT=${OUTPUT_FORMAT:-dir}
 
 set -e
-set -x
 
-IMAGE_NAME="$1"
+if [ -z "$IMAGE_NAME" ]; then
+    USE_DOCKERFILE=true
+    if [ -z "$DOCKER_FILE" -a -f 'Dockerfile' ]; then
+        DOCKER_FILE='./'
+    elif [ ! -f "$DOCKER_FILE" ]; then
+	echo "Could not find specific docker file: $DOCKER_FILE"
+	exit 1
+    fi
+else
+    USE_DOCKERFILE=false
+fi
+
+case "$OUTPUT_FORMAT" in
+    dir )
+        ;;
+    cpio )
+        ;;
+    ext2 )
+        echo "EXT2 format output is not directly supported by this program.";
+        exit 1
+        ;;
+    * )
+        echo "Unsupported output format: '$OUTPUT_FORMAT'.";
+        print_usage
+        ;;
+esac
+
+if [ -z $OUTPUT_NAME ]; then
+    echo "You must specify an output file name with the -o option."
+    echo
+    print_usage
+    exit 1
+fi
+
+if [ -d $OUTPUT_NAME -o -f $OUTPUT_NAME ]; then
+    echo "Output target '$OUTPUT_NAME' will be overwritten by this operation! Are you sure?"
+    select yn in "Yes" "No"; do
+	case $yn in
+	    Yes ) DELETE_OUTPUT=true ; break ;;
+	    No  ) echo "Aborting $(basename $0)"; exit 0 ;;
+	esac
+    done
+fi
 
-# Download the docker image from dockerhub, or build it locally
-# from a docker file.
-# If no command line option is provided, it assumes there is a local
-# docke file named exactly as "Dockerfile" in the current directory.
+# Download the docker image from dockerhub, or build it locally from a docker file.
 get_image()
 {
-    if [ -z $IMAGE_NAME ]; then
-        IMAGE_NAME="Dockerfile"
-    fi
-
-    if [ -f "$IMAGE_NAME" ]; then
-        DOCKER_IMG_NAME="temp-docker-image"
-        DOCKER_FILE=$IMAGE_NAME
-        IMAGE_NAME=$DOCKER_IMG_NAME
-        docker build -t $DOCKER_IMG_NAME . -f "$DOCKER_FILE"
+    if [ $USE_DOCKERFILE ]; then
+	IMAGE_NAME=$(docker build -q $DOCKER_FILE)
+	DELETE_IMAGE=true
     else
         docker pull $IMAGE_NAME
+	DELETE_IMAGE=false
     fi
 }
 
 # export a flattened copy of the container
-# note that we have to start up the image to flatten it
-# TODO: there's probably a fancier way to do this with more time
 export_image()
 {
-    docker run -d --name temp-extract $IMAGE_NAME
-    docker stop temp-extract
+    TEMP_FILE=$(mktemp)
+    TEMP_DIR=$(mktemp -d)
 
-    docker export temp-extract -o temp-extract.tar
-    sudo chmod 777 temp-extract.tar
-    docker rm temp-extract
+    # note that we have to start up the image to flatten it
+    TEMP_INAME=$(docker run -d $IMAGE_NAME)
+    docker stop $TEMP_INAME >/dev/null
+    docker export $TEMP_INAME -o $TEMP_FILE
+    docker rm $TEMP_INAME >/dev/null
+    if [ -z $DELETE_IMAGE ]; then
+	docker rmi $IMAGE_NAME
+    fi
 
-    rm -rf appdir && mkdir appdir
-    tar xf temp-extract.tar -C appdir
-}
+    # finally, create the output cpio object from the docker export
+    tar xf $TEMP_FILE -C $TEMP_DIR
```

Comment:
Can we take care of the name server issue at this point?

Reply:
I think that can be addressed in a separate PR, as it's a separate issue.